feat: support retrying

ShamrockLee opened the pull request:
### Description

This pull request adds a parameter `retries` to methods `Dials.list` and `Dials.list_all`, backed by `urllib3.util.Retry` and `requests.adapters.HTTPAdapter`.

Long downloads are prone to network issues, which can be mitigated by adding retrying functionality. This PR forwards the `max_retries` parameter of `HTTPAdapter` as `retries` to `Dials.list` and `Dials.list_all` and provides the retrying functionality without additional dependencies.

![Screenshot_20240813_110549](https://github.com/user-attachments/assets/c9c739ca-e5b7-49c3-a381-2330d310997c)

### Progress

I'm testing the functionality by downloading 1D histograms of four MEs from `/JetMET0/Run2024D-PromptReco-v1/DQMIO`.

~How do I add and run the tests inside the source tree?~ Do you think I should also update the documentation?

Update: I now see the instructions in the README.md.

ShamrockLee committed:
feat: BaseAuthorizedAPIClient.list[_all]: support retries

Add a parameter `retries` to methods `list` and `list_all` using
`BaseAPIClient._requests_get`.

gabrielmscampos commented:
If you are part of the cms-dqm-coreTeam e-group you can find the SECRET_KEY mentioned in the documentation from any api-client application in CERN's Applications Portal. DM me on Mattermost if you are not a member, so I can create an api client id/secret for you.

ShamrockLee commented:
@gabrielmscampos Thank you for helping out! I just sent you a direct message on Mattermost a minute ago.

ShamrockLee committed:
BaseAPIClient: add _requests_get_retriable

Add helper function BaseAPIClient._requests_get_retriable that accepts
an additional argument `retries` and utilize the retry functionality
provided by `requests.adapters.HTTPAdapter` and `urllib3.util.Retry`.

ShamrockLee committed:
tests: specify retries

ShamrockLee commented:
Sorry for taking so long to respond.

The test `test_list_all_h2d` succeeded after specifying `retries=Retry(total=5, backoff_factor=0.1)`. However, the test `test_get_h2d` keeps failing regardless of whether this change is applied, complaining of `FAILED tests/integration/test_h2d_client.py::test_get_h2d - requests.exceptions.HTTPError: 404 Client Error: Not Found for url...`.

ShamrockLee commented:
By the way, since the test cases use `Retry` from `urllib3.util` in this PR, and the retrying would be more useful when used together with `urllib3.util.Retry`, should we include `urllib3` as a (direct) dependency of this project?

`urllib3` is already an indirect dependency of this project because `requests` builds upon it.

gabrielmscampos commented:
Hi @ShamrockLee, sorry for the gelay. I was working in another DQM-DC project in the meantime and now I'm back on DIALS.

You are right, the following test `tests/integration/test_h2d_client.py::test_get_h2d` is failing. I inspected the database and notice that the `me_id=133` doesn't exist. I guess it existed at some point and it was wiped out during a full database re-ingestion. You can ignore this test or simply update `me_id=100` and the test should pass.

I agree with you that `urllib3` should become a direct dependency.

gabrielmscampos committed:
fix: update `test_get_h2d` test parameter `me_id` to `100`.

- 133 does not exist anymore in the database

gabrielmscampos committed:
chore: require urllib3 as direct dependency

gabrielmscampos commented:
I updated your PR to fix the aforementioned test and to add `urllib3` as direct dependency in `pyproject.toml`. All tests passed:

```
================================================================================= test session starts ==================================================================================
platform linux -- Python 3.9.19, pytest-8.2.0, pluggy-1.5.0
rootdir: ...
configfile: pyproject.toml
plugins: anyio-4.3.0
collected 24 items                                                                                                                                                                     

tests/integration/test_auth_client.py ...                                                                                                                                        [ 12%]
tests/integration/test_file_index_client.py ...                                                                                                                                  [ 25%]
tests/integration/test_h1d_client.py ...                                                                                                                                         [ 37%]
tests/integration/test_h2d_client.py ...                                                                                                                                         [ 50%]
tests/integration/test_lumisection_client.py ...                                                                                                                                 [ 62%]
tests/integration/test_mes_client.py ...                                                                                                                                         [ 75%]
tests/integration/test_oms_proxy_client.py ...                                                                                                                                   [ 87%]
tests/integration/test_run_client.py ...                                                                                                                                         [100%]

============================================================================ 24 passed in 90.86s (0:01:30) =============================================================================
```

gabrielmscampos commented:
Very cool solution, I think we can add it to the `get` method as well. The problem is that I'm overwriting the `get` method in almost all others clients because each one has a different `endpoint_url` depending on the inputs.

For example, the `h2d` client:

```python
    def get(self, dataset_id: int, run_number: int, ls_number: int):
        endpoint_url = f"{self.api_url}{self.lookup_url}{dataset_id}/{run_number}/{ls_number}/"
```

and the `run` client:

```python
    def get(self, dataset_id: int, run_number: int):
        endpoint_url = f"{self.api_url}{self.lookup_url}{dataset_id}/{run_number}/"
```

I think I can work around it by generalizing the base `get` from `BaseAuthorizedAPIClient` into something like this:

```python
    def get(self, *args):
        # Construct the endpoint URL based on the arguments
        endpoint_url = self.api_url + self.lookup_url + "/".join(map(str, args)) + "/"
```

I'll leave that for another PR.

Thank you very much!

gabrielmscampos committed:
docs: update docs to include retry instructions
